### PR TITLE
Remove `DAGResponseWriter::finishWrite`

### LINKED ISSUE #6355: random failpoint test faild for tiflash crash
## Bug Report

Please answer these questions before submitting your issue. Thanks!

### 1. Minimal reproduce step (Required)

```
[2022/11/23 10:10:42.938 +08:00] [ERROR] [BaseDaemon.cpp:377] [########################################] [source=BaseDaemon] [thread_id=2125]
[2022/11/23 10:10:42.951 +08:00] [ERROR] [BaseDaemon.cpp:378] ["(from thread 2056) Received signal Aborted(6)."] [source=BaseDaemon] [thread_id=2125]
[2022/11/23 10:10:43.117 +08:00] [ERROR] [BaseDaemon.cpp:570] ["\n       0x9240fbd\tfaultSignalHandler(int, siginfo_t*, void*) [tiflash+153358269]\n                \tlibs/libdaemon/src/BaseDaemon.cpp:221\n  0x7fe3c3674630\t<unknown symbol> [libpthread.so.0+63024]\n  0x7fe3c29a9387\tgsignal [libc.so.6+222087]\n  0x7fe3c29aaa78\t__GI_abort [libc.so.6+227960]\n  0x7fe3c29a21a6\t__assert_fail_base [libc.so.6+192934]\n  0x7fe3c29a2252\t__assert_fail [libc.so.6+193106]\n       0x3985972\tDB::StreamingDAGResponseWriter<std::__1::shared_ptr<DB::MPPTunnelSet> >::finishWrite() [tiflash+60316018]\n                \tdbms/src/Flash/Coprocessor/StreamingDAGResponseWriter.cpp:66\n       0xa347922\tDB::ExchangeSenderBlockInputStream::readSuffixImpl() [tiflash+171211042]\n                \tdbms/src/DataStreams/ExchangeSenderBlockInputStream.h:49\n       0x926d762\tDB::IProfilingBlockInputStream::readSuffix() [tiflash+153540450]\n                \tdbms/src/DataStreams/IProfilingBlockInputStream.cpp:141\n       0xa34b8ce\tDB::UnionBlockInputStream<(DB::StreamUnionMode)0, true>::readSuffix() [tiflash+171227342]\n                \tdbms/src/DataStreams/UnionBlockInputStream.h:255\n       0xa394724\tDB::MPPTask::runImpl() [tiflash+171525924]\n                \tdbms/src/Flash/Mpp/MPPTask.cpp:382\n       0xa39834d\tDB::MPPTask::run()::$_12::operator()() const [tiflash+171541325]\n                \tdbms/src/Flash/Mpp/MPPTask.cpp:117\n       0xa398315\tdecltype(static_cast<DB::MPPTask::run()::$_12&>(fp)()) std::__1::__invoke<DB::MPPTask::run()::$_12&>(DB::MPPTask::run()::$_12&) [tiflash+171541269]\n                \t/usr/local/bin/../include/c++/v1/type_traits:3918\n       0xa3982cd\tvoid std::__1::__invoke_void_return_wrapper<void, true>::__call<DB::MPPTask::run()::$_12&>(DB::MPPTask::run()::$_12&) [tiflash+171541197]\n                \t/usr/local/bin/../include/c++/v1/__functional/invoke.h:61\n       0xa39829d\tstd::__1::__function::__alloc_func<DB::MPPTask::run()::$_12, std::__1::allocator<DB::MPPTask::run()::$_12>, void ()>::operator()() [tiflash+171541149]\n                \t/usr/local/bin/../include/c++/v1/__functional/function.h:171\n       0xa3973c9\tstd::__1::__function::__func<DB::MPPTask::run()::$_12, std::__1::allocator<DB::MPPTask::run()::$_12>, void ()>::operator()() [tiflash+171537353]\n                \t/usr/local/bin/../include/c++/v1/__functional/function.h:345\n       0x3b79246\tstd::__1::__function::__value_func<void ()>::operator()() const [tiflash+62362182]\n                \t/usr/local/bin/../include/c++/v1/__functional/function.h:498\n       0x3b791f5\tstd::__1::function<void ()>::operator()() const [tiflash+62362101]\n                \t/usr/local/bin/../include/c++/v1/__functional/function.h:1175\n       0x3b85c75\tdecltype(static_cast<std::__1::function<void ()>>(fp)()) std::__1::__invoke_constexpr<std::__1::function<void ()> >(std::__1::function<void ()>&&) [tiflash+62413941]\n                \t/usr/local/bin/../include/c++/v1/type_traits:3924\n       0x3b85c51\tdecltype(auto) std::__1::__apply_tuple_impl<std::__1::function<void ()>, std::__1::tuple<> >(std::__1::function<void ()>&&, std::__1::tuple<>&&, std::__1::__tuple_indices<>) [tiflash+62413905]\n                \t/usr/local/bin/../include/c++/v1/tuple:1536\n       0x3b85c21\tdecltype(auto) std::__1::apply<std::__1::function<void ()>, std::__1::tuple<> >(std::__1::function<void ()>&&, std::__1::tuple<>&&) [tiflash+62413857]\n                \t/usr/local/bin/../include/c++/v1/tuple:1545\n       0x3b85bbc\tauto DB::wrapInvocable<std::__1::function<void ()> >(bool, std::__1::function<void ()>&&)::'lambda'()::operator()() [tiflash+62413756]\n                \tdbms/src/Common/wrapInvocable.h:36\n       0x3b866c9\tDB::ExecutableTask<auto DB::wrapInvocable<std::__1::function<void ()> >(bool, std::__1::function<void ()>&&)::'lambda'()>::execute() [tiflash+62416585]\n                \tdbms/src/Common/ExecutableTask.h:52\n       0x3b8b15d\tDB::DynamicThreadPool::executeTask(std::__1::unique_ptr<DB::IExecutableTask, std::__1::default_delete<DB::IExecutableTask> >&) [tiflash+62435677]\n                \tdbms/src/Common/DynamicThreadPool.cpp:101\n       0x3b8b023\tDB::DynamicThreadPool::dynamicWork(std::__1::unique_ptr<DB::IExecutableTask, std::__1::default_delete<DB::IExecutableTask> >) [tiflash+62435363]\n                \tdbms/src/Common/DynamicThreadPool.cpp:142\n       0x3b8f750\tdecltype(*(static_cast<DB::DynamicThreadPool*>(fp0)).*fp(static_cast<std::__1::unique_ptr<DB::IExecutableTask, std::__1::default_delete<DB::IExecutableTask> >>(fp1))) std::__1::__invoke<void (DB::DynamicThreadPool::* const&)(std::__1::unique_ptr<DB::IExecutableTask, std::__1::default_delete<DB::IExecutableTask> >), DB::DynamicThreadPool*, std::__1::unique_ptr<DB::IExecutableTask, std::__1::default_delete<DB::IExecutableTask> >, void>(void (DB::DynamicThreadPool::* const&)(std::__1::unique_ptr<DB::IExecutableTask, std::__1::default_delete<DB::IExecutableTask> >), DB::DynamicThreadPool*&&, std::__1::unique_ptr<DB::IExecutableTask, std::__1::default_delete<DB::IExecutableTask> >&&) [tiflash+62453584]\n                \t/usr/local/bin/../include/c++/v1/type_traits:3859\n       0x3b8f676\tstd::__1::invoke_result<void (DB::DynamicThreadPool::* const&)(std::__1::unique_ptr<DB::IExecutableTask, std::__1::default_delete<DB::IExecutableTask> >), DB::DynamicThreadPool*, std::__1::unique_ptr<DB::IExecutableTask, std::__1::default_delete<DB::IExecutableTask> > >::type std::__1::invoke<void (DB::DynamicThreadPool::* const&)(std::__1::unique_ptr<DB::IExecutableTask, std::__1::default_delete<DB::IExecutableTask> >), DB::DynamicThreadPool*, std::__1::unique_ptr<DB::IExecutableTask, std::__1::default_delete<DB::IExecutableTask> > >(void (DB::DynamicThreadPool::* const&)(std::__1::unique_ptr<DB::IExecutableTask, std::__1::default_delete<DB::IExecutableTask> >), DB::DynamicThreadPool*&&, std::__1::unique_ptr<DB::IExecutableTask, std::__1::default_delete<DB::IExecutableTask> >&&) [tiflash+62453366]\n                \t/usr/local/bin/../include/c++/v1/__functional/invoke.h:93\n       0x3b8f5b5\tauto std::__1::thread DB::ThreadFactory::newThread<void (DB::DynamicThreadPool::*)(std::__1::unique_ptr<DB::IExecutableTask, std::__1::default_delete<DB::IExecutableTask> >), DB::DynamicThreadPool*, std::__1::unique_ptr<DB::IExecutableTask, std::__1::default_delete<DB::IExecutableTask> > >(bool, std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> >, void (DB::DynamicThreadPool::*&&)(std::__1::unique_ptr<DB::IExecutableTask, std::__1::default_delete<DB::IExecutableTask> >), DB::DynamicThreadPool*&&, std::__1::unique_ptr<DB::IExecutableTask, std::__1::default_delete<DB::IExecutableTask> >&&)::'lambda'(auto&&...)::operator()<DB::DynamicThreadPool*, std::__1::unique_ptr<DB::IExecutableTask, std::__1::default_delete<DB::IExecutableTask> > >(auto&&...) const [tiflash+62453173]\n                \tdbms/src/Common/ThreadFactory.h:47\n       0x3b8f3f5\tdecltype(static_cast<void (DB::DynamicThreadPool::*>(fp)(static_cast<DB::DynamicThreadPool*>(fp0), static_cast<std::__1::unique_ptr<DB::IExecutableTask, std::__1::default_delete<DB::IExecutableTask> >>(fp0))) std::__1::__invoke<std::__1::thread DB::ThreadFactory::newThread<void (DB::DynamicThreadPool::*)(std::__1::unique_ptr<DB::IExecutableTask, std::__1::default_delete<DB::IExecutableTask> >), DB::DynamicThreadPool*, std::__1::unique_ptr<DB::IExecutableTask, std::__1::default_delete<DB::IExecutableTask> > >(bool, std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> >, void (DB::DynamicThreadPool::*&&)(std::__1::unique_ptr<DB::IExecutableTask, std::__1::default_delete<DB::IExecutableTask> >), DB::DynamicThreadPool*&&, std::__1::unique_ptr<DB::IExecutableTask, std::__1::default_delete<DB::IExecutableTask> >&&)::'lambda'(auto&&...), DB::DynamicThreadPool*, std::__1::unique_ptr<DB::IExecutableTask, std::__1::default_delete<DB::IExecutableTask> > >(void (DB::DynamicThreadPool::*&&)(std::__1::unique_ptr<DB::IExecutableTask, std::__1::default_delete<DB::IExecutableTask> >), DB::DynamicThreadPool*&&, std::__1::unique_ptr<DB::IExecutableTask, std::__1::default_delete<DB::IExecutableTask> >&&) [tiflash+62452725]\n                \t/usr/local/bin/../include/c++/v1/type_traits:3918\n       0x3b8f3c6\tvoid std::__1::__thread_execute<std::__1::unique_ptr<std::__1::__thread_struct, std::__1::default_delete<std::__1::__thread_struct> >, std::__1::thread DB::ThreadFactory::newThread<void (DB::DynamicThreadPool::*)(std::__1::unique_ptr<DB::IExecutableTask, std::__1::default_delete<DB::IExecutableTask> >), DB::DynamicThreadPool*, std::__1::unique_ptr<DB::IExecutableTask, std::__1::default_delete<DB::IExecutableTask> > >(bool, std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> >, void (DB::DynamicThreadPool::*&&)(std::__1::unique_ptr<DB::IExecutableTask, std::__1::default_delete<DB::IExecutableTask> >), DB::DynamicThreadPool*&&, std::__1::unique_ptr<DB::IExecutableTask, std::__1::default_delete<DB::IExecutableTask> >&&)::'lambda'(auto&&...), DB::DynamicThreadPool*, std::__1::unique_ptr<DB::IExecutableTask, std::__1::default_delete<DB::IExecutableTask> >, 2ul, 3ul>(std::__1::tuple<void (DB::DynamicThreadPool::*)(std::__1::unique_ptr<DB::IExecutableTask, std::__1::default_delete<DB::IExecutableTask> >), std::__1::thread DB::ThreadFactory::newThread<void (DB::DynamicThreadPool::*)(std::__1::unique_ptr<DB::IExecutableTask, std::__1::default_delete<DB::IExecutableTask> >), DB::DynamicThreadPool*, std::__1::unique_ptr<DB::IExecutableTask, std::__1::default_delete<DB::IExecutableTask> > >(bool, std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> >, void (DB::DynamicThreadPool::*&&)(std::__1::unique_ptr<DB::IExecutableTask, std::__1::default_delete<DB::IExecutableTask> >), DB::DynamicThreadPool*&&, std::__1::unique_ptr<DB::IExecutableTask, std::__1::default_delete<DB::IExecutableTask> >&&)::'lambda'(auto&&...), DB::DynamicThreadPool*, std::__1::unique_ptr<DB::IExecutableTask, std::__1::default_delete<DB::IExecutableTask> > >&, std::__1::__tuple_indices<2ul, 3ul>) [tiflash+62452678]\n                \t/usr/local/bin/../include/c++/v1/thread:280\n       0x3b8f05b\tvoid* std::__1::__thread_proxy<std::__1::tuple<std::__1::unique_ptr<std::__1::__thread_struct, std::__1::default_delete<std::__1::__thread_struct> >, std::__1::thread DB::ThreadFactory::newThread<void (DB::DynamicThreadPool::*)(std::__1::unique_ptr<DB::IExecutableTask, std::__1::default_delete<DB::IExecutableTask> >), DB::DynamicThreadPool*, std::__1::unique_ptr<DB::IExecutableTask, std::__1::default_delete<DB::IExecutableTask> > >(bool, std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> >, void (DB::DynamicThreadPool::*&&)(std::__1::unique_ptr<DB::IExecutableTask, std::__1::default_delete<DB::IExecutableTask> >), DB::DynamicThreadPool*&&, std::__1::unique_ptr<DB::IExecutableTask, std::__1::default_delete<DB::IExecutableTask> >&&)::'lambda'(auto&&...), DB::DynamicThreadPool*, std::__1::unique_ptr<DB::IExecutableTask, std::__1::default_delete<DB::IExecutableTask> > > >(void*) [tiflash+62451803]\n                \t/usr/local/bin/../include/c++/v1/thread:291\n  0x7fe3c366cea5\tstart_thread [libpthread.so.0+32421]"] [source=BaseDaemon] [thread_id=2125]
```
### 4. What is your TiFlash version? (Required)
master

### PR DESCRIPTION
**Pull request by SeaRise**
### What problem does this PR solve?

Issue Number: close #6355

Problem Summary:

### What is changed and how it works?

- remove uesless check `assert(rows_in_blocks == 0)` in ExchangeWriter and StreamingWriter
- move the almost of code in `UnaryDAGResponseWriter::finishWrite` to `UnaryDAGResponseWriter::flush`
- move filling execution summary from `UnaryDAGResponseWriter::finishWrite` to `DAGDriver::execute`

The reason for the crash is
1. One of the MPPTask occur exception and call `manager->abortMPPQuery(id.start_ts, updated_msg, AbortType::ONERROR)`
2. Other mpp task is called `MPPTask::abortDataStreams(AbortType::ONERROR)`
3. And then the final UnionBlockInputStream return an empty block in advance and finish `Stream->read`, and `StreamingWriter::flush` is not called, so `rows_in_blocks != 0`
4. the stream->readPrefix is called, and the tiflash node crash in `StreamingDAGResponseWriter::finishWrite() { assert(rows_in_blocks == 0) }`...

### Check List

Tests 

- [ ] Unit test
- [ ] Integration test
- [ ] Manual test (add detailed scripts or steps below)
- [ ] No code

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

Documentation

- [ ] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features
- [ ] Changes MySQL compatibility

### Release note

```release-note
None
```

### CODE REVIEW

File: dbms/src/Flash/Coprocessor/DAGDriver.cpp [hunk 1]
```diff
@@ -110,6 +110,11 @@ try
             dag_context);
         dag_output_stream = std::make_shared<DAGBlockOutputStream>(streams.in->getHeader(), std::move(response_writer));
         copyData(*streams.in, *dag_output_stream);
+        if (dag_context.collect_execution_summaries)
```
Comment (bestwoody) [inline dbms/src/Flash/Coprocessor/DAGDriver.cpp:113]: does it have the same effect as the original one in `UnaryDAGResponseWriter::finishWrite`

Comment (SeaRise) [reply dbms/src/Flash/Coprocessor/DAGDriver.cpp:113]: yes, it does.

**Review by windtalker — APPROVED** [review]
LGTM

**Review by ywqzzy — APPROVED** [review]

**Pull request merged**